In [2]:
#importing dependencies
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import warnings
warnings.filterwarnings("ignore")

In [3]:
#reading data from city_day.csv taken from kaggle from 01-01-2015 to 31-12-2024 of various Indian Cities
df = pd.read_csv("India_Cities_AQI_Weather_2015_2024_Combined.csv")

In [4]:
df.head()

Date   City  State  AQI AQI_Category  PM2.5  PM10  NO2  SO2    CO  \
0  2015-01-01  Delhi  Delhi  282         Poor    121   180   86   63  5.75   
1  2015-01-02  Delhi  Delhi  210         Poor     85   139   59   45  4.10   
2  2015-01-03  Delhi  Delhi  257         Poor    107   168   72   59  5.54   
3  2015-01-04  Delhi  Delhi  193     Moderate     88   126   48   36  3.58   
4  2015-01-05  Delhi  Delhi  274         Poor    124   145   83   53  4.85   

   ...  Min_Temperature_C  Avg_Temperature_C  Humidity_Percent  Rainfall_mm  \
0  ...               -6.0               -1.5              76.4          0.0   
1  ...               -8.8               -3.1              48.9          0.0   
2  ...               -4.5               -0.3              53.7          0.0   
3  ...               -7.3               -0.6              76.7          0.0   
4  ...               -3.1                4.4              66.1          0.0   

   Wind_Speed_kmh  Wind_Direction  Atmospheric_Pressure_hPa  Visibility_km  \
0            11.1               N                   1010.55           14.9   
1             9.2               S                   1015.14           14.5   
2             7.1              NE                   1017.66           14.2   
3            10.4              NW                   1009.99           15.0   
4            12.8              NE                   1016.24           15.0   

  Month  Season  
0     1  Winter  
1     1  Winter  
2     1  Winter  
3     1  Winter  
4     1  Winter  

[5 rows x 23 columns]

### Analysing data 

#### Understanding the Data and handling missing values

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18265 entries, 0 to 18264
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Date                      18265 non-null  object 
 1   City                      18265 non-null  object 
 2   State                     18265 non-null  object 
 3   AQI                       18265 non-null  int64  
 4   AQI_Category              18265 non-null  object 
 5   PM2.5                     18265 non-null  int64  
 6   PM10                      18265 non-null  int64  
 7   NO2                       18265 non-null  int64  
 8   SO2                       18265 non-null  int64  
 9   CO                        18265 non-null  float64
 10  O3                        18265 non-null  int64  
 11  NH3                       18265 non-null  int64  
 12  Max_Temperature_C         18265 non-null  float64
 13  Min_Temperature_C         18265 non-null  float64
 14  Avg_Te

In [6]:
df.shape

(18265, 23)

In [7]:
#checking null values
df.isnull().sum()

Date                        0
City                        0
State                       0
AQI                         0
AQI_Category                0
PM2.5                       0
PM10                        0
NO2                         0
SO2                         0
CO                          0
O3                          0
NH3                         0
Max_Temperature_C           0
Min_Temperature_C           0
Avg_Temperature_C           0
Humidity_Percent            0
Rainfall_mm                 0
Wind_Speed_kmh              0
Wind_Direction              0
Atmospheric_Pressure_hPa    0
Visibility_km               0
Month                       0
Season                      0
dtype: int64

In [8]:
df.columns

Index(['Date', 'City', 'State', 'AQI', 'AQI_Category', 'PM2.5', 'PM10', 'NO2',
       'SO2', 'CO', 'O3', 'NH3', 'Max_Temperature_C', 'Min_Temperature_C',
       'Avg_Temperature_C', 'Humidity_Percent', 'Rainfall_mm',
       'Wind_Speed_kmh', 'Wind_Direction', 'Atmospheric_Pressure_hPa',
       'Visibility_km', 'Month', 'Season'],
      dtype='object')

In [9]:
df.describe()

AQI         PM2.5          PM10           NO2           SO2  \
count  18265.000000  18265.000000  18265.000000  18265.000000  18265.000000   
mean     117.620476     46.519244     70.035532     34.795292     22.990857   
std       50.482348     22.559589     33.684170     17.062194     11.242799   
min       21.000000      1.000000      1.000000      1.000000      1.000000   
25%       81.000000     30.000000     46.000000     23.000000     15.000000   
50%      107.000000     43.000000     65.000000     33.000000     21.000000   
75%      143.000000     59.000000     89.000000     44.000000     29.000000   
max      371.000000    167.000000    242.000000    118.000000     77.000000   

                 CO            O3           NH3  Max_Temperature_C  \
count  18265.000000  18265.000000  18265.000000       18265.000000   
mean       2.357313     40.670025     17.231371          25.974563   
std        1.126961     21.112863      9.518506           9.786889   
min        0.100000      1.000000      1.000000          -3.700000   
25%        1.570000     26.000000     10.000000          18.800000   
50%        2.200000     38.000000     16.000000          26.100000   
75%        2.980000     53.000000     23.000000          33.500000   
max        8.960000    142.000000     65.000000          52.100000   

       Min_Temperature_C  Avg_Temperature_C  Humidity_Percent   Rainfall_mm  \
count       18265.000000       18265.000000      18265.000000  18265.000000   
mean           14.468119          20.221276         76.384216      1.266685   
std             9.987360           9.836193         13.852881      5.116442   
min           -17.800000         -10.800000         30.000000      0.000000   
25%             7.100000          13.000000         66.600000      0.000000   
50%            14.700000          20.300000         76.700000      0.000000   
75%            22.100000          27.800000         86.700000      0.000000   
max            43.100000          47.600000        100.000000    134.400000   

       Wind_Speed_kmh  Atmospheric_Pressure_hPa  Visibility_km         Month  
count    18265.000000              18265.000000   18265.000000  18265.000000  
mean         8.822502               1013.242090       9.774481      6.522310  
std          3.592870                 10.039602       2.746346      3.448967  
min          0.000000                962.360000       0.500000      1.000000  
25%          6.300000               1006.500000       7.900000      4.000000  
50%          8.800000               1013.210000       9.800000      7.000000  
75%         11.200000               1020.090000      11.800000     10.000000  
max         23.200000               1055.000000      15.000000     12.000000

In [10]:
print(df["City"].value_counts())

City
Delhi        3653
Chennai      3653
Kolkata      3653
Mumbai       3653
Bengaluru    3653
Name: count, dtype: int64


#### Observation

This dataset contains daily amount of pollutant present in air in top 5 cities of India - Delhi, Mumbai, Chennai, Kolkata, and Banglore.
The dataset has 18265 rows and 16 columns out of which three columns are categorical (City, Date, AQI_Bucket) rest are numerical amount of each pollutant present in air. 

In [11]:
#converting date column into datetime object
df["Date"]=pd.to_datetime(df["Date"])

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18265 entries, 0 to 18264
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Date                      18265 non-null  datetime64[ns]
 1   City                      18265 non-null  object        
 2   State                     18265 non-null  object        
 3   AQI                       18265 non-null  int64         
 4   AQI_Category              18265 non-null  object        
 5   PM2.5                     18265 non-null  int64         
 6   PM10                      18265 non-null  int64         
 7   NO2                       18265 non-null  int64         
 8   SO2                       18265 non-null  int64         
 9   CO                        18265 non-null  float64       
 10  O3                        18265 non-null  int64         
 11  NH3                       18265 non-null  int64         
 12  Max_Temperature_C 

In [13]:
#diving group based on city names
group_df = df.groupby("City")

#making city wise dataframe
delhi_df = group_df.get_group("Delhi")

In [14]:
delhi_df

Date   City  State  AQI AQI_Category  PM2.5  PM10  NO2  SO2    CO  \
0    2015-01-01  Delhi  Delhi  282         Poor    121   180   86   63  5.75   
1    2015-01-02  Delhi  Delhi  210         Poor     85   139   59   45  4.10   
2    2015-01-03  Delhi  Delhi  257         Poor    107   168   72   59  5.54   
3    2015-01-04  Delhi  Delhi  193     Moderate     88   126   48   36  3.58   
4    2015-01-05  Delhi  Delhi  274         Poor    124   145   83   53  4.85   
...         ...    ...    ...  ...          ...    ...   ...  ...  ...   ...   
3648 2024-12-27  Delhi  Delhi  249         Poor     99   155   78   46  4.08   
3649 2024-12-28  Delhi  Delhi  232         Poor     87   119   63   49  5.02   
3650 2024-12-29  Delhi  Delhi  198     Moderate     81   128   68   43  4.37   
3651 2024-12-30  Delhi  Delhi  199     Moderate     88   129   48   41  3.76   
3652 2024-12-31  Delhi  Delhi  292         Poor    119   202   98   54  6.00   

      ...  Min_Temperature_C  Avg_Temperature_C  Humidity_Percent  \
0     ...               -6.0               -1.5              76.4   
1     ...               -8.8               -3.1              48.9   
2     ...               -4.5               -0.3              53.7   
3     ...               -7.3               -0.6              76.7   
4     ...               -3.1                4.4              66.1   
...   ...                ...                ...               ...   
3648  ...               -5.9               -1.0              58.5   
3649  ...               -6.7               -2.3              68.2   
3650  ...              -10.2               -4.6              67.9   
3651  ...               -5.9                0.1              55.9   
3652  ...               -9.2               -2.6              77.2   

      Rainfall_mm  Wind_Speed_kmh  Wind_Direction  Atmospheric_Pressure_hPa  \
0             0.0            11.1               N                   1010.55   
1             0.0             9.2               S                   1015.14   
2             0.0             7.1              NE                   1017.66   
3             0.0            10.4              NW                   1009.99   
4             0.0            12.8              NE                   1016.24   
...           ...             ...             ...                       ...   
3648          0.0             7.0              NW                    993.60   
3649          0.0            10.1               S                   1033.39   
3650          0.0             9.2              NE                   1011.49   
3651          0.0             9.2               E                   1011.07   
3652          0.0            13.2               S                   1019.73   

      Visibility_km Month  Season  
0              14.9     1  Winter  
1              14.5     1  Winter  
2              14.2     1  Winter  
3              15.0     1  Winter  
4              15.0     1  Winter  
...             ...   ...     ...  
3648           14.5    12  Winter  
3649           14.3    12  Winter  
3650           14.3    12  Winter  
3651           14.1    12  Winter  
3652           10.9    12  Winter  

[3653 rows x 23 columns]

In [17]:
#reseting indices
delhi_df = delhi_df.reset_index(drop=True)

In [18]:
delhi_df

Date   City  State  AQI AQI_Category  PM2.5  PM10  NO2  SO2    CO  \
0    2015-01-01  Delhi  Delhi  282         Poor    121   180   86   63  5.75   
1    2015-01-02  Delhi  Delhi  210         Poor     85   139   59   45  4.10   
2    2015-01-03  Delhi  Delhi  257         Poor    107   168   72   59  5.54   
3    2015-01-04  Delhi  Delhi  193     Moderate     88   126   48   36  3.58   
4    2015-01-05  Delhi  Delhi  274         Poor    124   145   83   53  4.85   
...         ...    ...    ...  ...          ...    ...   ...  ...  ...   ...   
3648 2024-12-27  Delhi  Delhi  249         Poor     99   155   78   46  4.08   
3649 2024-12-28  Delhi  Delhi  232         Poor     87   119   63   49  5.02   
3650 2024-12-29  Delhi  Delhi  198     Moderate     81   128   68   43  4.37   
3651 2024-12-30  Delhi  Delhi  199     Moderate     88   129   48   41  3.76   
3652 2024-12-31  Delhi  Delhi  292         Poor    119   202   98   54  6.00   

      ...  Min_Temperature_C  Avg_Temperature_C  Humidity_Percent  \
0     ...               -6.0               -1.5              76.4   
1     ...               -8.8               -3.1              48.9   
2     ...               -4.5               -0.3              53.7   
3     ...               -7.3               -0.6              76.7   
4     ...               -3.1                4.4              66.1   
...   ...                ...                ...               ...   
3648  ...               -5.9               -1.0              58.5   
3649  ...               -6.7               -2.3              68.2   
3650  ...              -10.2               -4.6              67.9   
3651  ...               -5.9                0.1              55.9   
3652  ...               -9.2               -2.6              77.2   

      Rainfall_mm  Wind_Speed_kmh  Wind_Direction  Atmospheric_Pressure_hPa  \
0             0.0            11.1               N                   1010.55   
1             0.0             9.2               S                   1015.14   
2             0.0             7.1              NE                   1017.66   
3             0.0            10.4              NW                   1009.99   
4             0.0            12.8              NE                   1016.24   
...           ...             ...             ...                       ...   
3648          0.0             7.0              NW                    993.60   
3649          0.0            10.1               S                   1033.39   
3650          0.0             9.2              NE                   1011.49   
3651          0.0             9.2               E                   1011.07   
3652          0.0            13.2               S                   1019.73   

      Visibility_km Month  Season  
0              14.9     1  Winter  
1              14.5     1  Winter  
2              14.2     1  Winter  
3              15.0     1  Winter  
4              15.0     1  Winter  
...             ...   ...     ...  
3648           14.5    12  Winter  
3649           14.3    12  Winter  
3650           14.3    12  Winter  
3651           14.1    12  Winter  
3652           10.9    12  Winter  

[3653 rows x 23 columns]

### Forecasting using FB Prophet

In [21]:
delhi_df

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


Date   City  State  AQI AQI_Category  PM2.5  PM10  NO2  SO2    CO  \
0    2015-01-01  Delhi  Delhi  282         Poor    121   180   86   63  5.75   
1    2015-01-02  Delhi  Delhi  210         Poor     85   139   59   45  4.10   
2    2015-01-03  Delhi  Delhi  257         Poor    107   168   72   59  5.54   
3    2015-01-04  Delhi  Delhi  193     Moderate     88   126   48   36  3.58   
4    2015-01-05  Delhi  Delhi  274         Poor    124   145   83   53  4.85   
...         ...    ...    ...  ...          ...    ...   ...  ...  ...   ...   
3648 2024-12-27  Delhi  Delhi  249         Poor     99   155   78   46  4.08   
3649 2024-12-28  Delhi  Delhi  232         Poor     87   119   63   49  5.02   
3650 2024-12-29  Delhi  Delhi  198     Moderate     81   128   68   43  4.37   
3651 2024-12-30  Delhi  Delhi  199     Moderate     88   129   48   41  3.76   
3652 2024-12-31  Delhi  Delhi  292         Poor    119   202   98   54  6.00   

      ...  Min_Temperature_C  Avg_Temperature_C  Humidity_Percent  \
0     ...               -6.0               -1.5              76.4   
1     ...               -8.8               -3.1              48.9   
2     ...               -4.5               -0.3              53.7   
3     ...               -7.3               -0.6              76.7   
4     ...               -3.1                4.4              66.1   
...   ...                ...                ...               ...   
3648  ...               -5.9               -1.0              58.5   
3649  ...               -6.7               -2.3              68.2   
3650  ...              -10.2               -4.6              67.9   
3651  ...               -5.9                0.1              55.9   
3652  ...               -9.2               -2.6              77.2   

      Rainfall_mm  Wind_Speed_kmh  Wind_Direction  Atmospheric_Pressure_hPa  \
0             0.0            11.1               N                   1010.55   
1             0.0             9.2               S                   1015.14   
2             0.0             7.1              NE                   1017.66   
3             0.0            10.4              NW                   1009.99   
4             0.0            12.8              NE                   1016.24   
...           ...             ...             ...                       ...   
3648          0.0             7.0              NW                    993.60   
3649          0.0            10.1               S                   1033.39   
3650          0.0             9.2              NE                   1011.49   
3651          0.0             9.2               E                   1011.07   
3652          0.0            13.2               S                   1019.73   

      Visibility_km Month  Season  
0              14.9     1  Winter  
1              14.5     1  Winter  
2              14.2     1  Winter  
3              15.0     1  Winter  
4              15.0     1  Winter  
...             ...   ...     ...  
3648           14.5    12  Winter  
3649           14.3    12  Winter  
3650           14.3    12  Winter  
3651           14.1    12  Winter  
3652           10.9    12  Winter  

[3653 rows x 23 columns]

In [22]:
delhi_df.columns

Index(['Date', 'City', 'State', 'AQI', 'AQI_Category', 'PM2.5', 'PM10', 'NO2',
       'SO2', 'CO', 'O3', 'NH3', 'Max_Temperature_C', 'Min_Temperature_C',
       'Avg_Temperature_C', 'Humidity_Percent', 'Rainfall_mm',
       'Wind_Speed_kmh', 'Wind_Direction', 'Atmospheric_Pressure_hPa',
       'Visibility_km', 'Month', 'Season'],
      dtype='object')

In [23]:
delhi_df.index

RangeIndex(start=0, stop=3653, step=1)

In [24]:
#splitting each pollutant
#pm2.5
delhi_pm25 = pd.DataFrame(delhi_df[["Date", "PM2.5"]])
delhi_pm25.columns = ["ds", "y"]

#pm10
delhi_pm10 = pd.DataFrame(delhi_df[["Date", "PM10"]])
delhi_pm10.columns = ["ds", "y"]

#NO2
delhi_no2 = pd.DataFrame(delhi_df[["Date", "NO2"]])
delhi_no2.columns = ["ds", "y"]

#NH3
delhi_nh3 = pd.DataFrame(delhi_df[["Date", "NH3"]])
delhi_nh3.columns = ["ds", "y"]

#CO
delhi_co = pd.DataFrame(delhi_df[["Date", "CO"]])
delhi_co.columns = ["ds", "y"]

#SO2
delhi_so2 = pd.DataFrame(delhi_df[["Date", "SO2"]])
delhi_so2.columns = ["ds", "y"]

#O3
delhi_o3 = pd.DataFrame(delhi_df[["Date", "O3"]])
delhi_o3.columns = ["ds", "y"]


In [25]:
delhi_pm25['ds']

0      2015-01-01
1      2015-01-02
2      2015-01-03
3      2015-01-04
4      2015-01-05
          ...    
3648   2024-12-27
3649   2024-12-28
3650   2024-12-29
3651   2024-12-30
3652   2024-12-31
Name: ds, Length: 3653, dtype: datetime64[ns]

In [26]:
from prophet import Prophet

In [27]:
#initialize prophet for each pollutant
model_pm25 = Prophet()
model_pm25.fit(delhi_pm25)

#future timestamps
future_pm25 = model_pm25.make_future_dataframe(periods=365)

#predictions
forecast_pm25 = model_pm25.predict(future_pm25)

#print results
print(forecast_pm25[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:23 - cmdstanpy - INFO - Chain [1] start processing
12:01:24 - cmdstanpy - INFO - Chain [1] done processing


             ds       yhat  yhat_lower  yhat_upper
4013 2025-12-27  83.845718   61.335735  107.351421
4014 2025-12-28  84.047201   61.637710  107.969904
4015 2025-12-29  93.251496   71.417840  116.128245
4016 2025-12-30  92.930717   69.344129  114.679079
4017 2025-12-31  91.848317   69.637954  115.792254


In [28]:
model_pm10 = Prophet()
model_pm10.fit(delhi_pm10)

#future timestamps
future_pm10 = model_pm10.make_future_dataframe(periods=365)

#predictions
forecast_pm10 = model_pm10.predict(future_pm10)

#print results
print(forecast_pm10[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:26 - cmdstanpy - INFO - Chain [1] start processing
12:01:27 - cmdstanpy - INFO - Chain [1] done processing


             ds        yhat  yhat_lower  yhat_upper
4013 2025-12-27  125.617609   92.410917  160.342919
4014 2025-12-28  127.109775   93.445596  163.688978
4015 2025-12-29  138.765564  104.957912  171.462855
4016 2025-12-30  138.322204  105.225691  171.947412
4017 2025-12-31  138.253702  103.482574  175.616867


In [29]:
#initialize prophet for each pollutant
model_no2 = Prophet()
model_no2.fit(delhi_no2)

#future timestamps
future_no2 = model_no2.make_future_dataframe(periods=365)

#predictions
forecast_no2 = model_no2.predict(future_no2)

#print results
print(forecast_no2[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:28 - cmdstanpy - INFO - Chain [1] start processing
12:01:29 - cmdstanpy - INFO - Chain [1] done processing


             ds       yhat  yhat_lower  yhat_upper
4013 2025-12-27  62.862590   45.813146   80.689518
4014 2025-12-28  63.239887   45.918049   80.426304
4015 2025-12-29  68.770745   51.839799   85.668752
4016 2025-12-30  68.817066   50.740330   86.043379
4017 2025-12-31  68.685106   51.322921   86.091318


In [30]:
#initialize prophet for each pollutant
model_nh3 = Prophet()
model_nh3.fit(delhi_nh3)

#future timestamps
future_nh3 = model_nh3.make_future_dataframe(periods=365)

#predictions
forecast_nh3 = model_nh3.predict(future_nh3)

#print results
print(forecast_nh3[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:30 - cmdstanpy - INFO - Chain [1] start processing
12:01:30 - cmdstanpy - INFO - Chain [1] done processing


             ds       yhat  yhat_lower  yhat_upper
4013 2025-12-27  31.079274   20.044805   41.538502
4014 2025-12-28  31.370148   20.885850   42.001698
4015 2025-12-29  34.326810   24.220248   45.705801
4016 2025-12-30  34.654162   23.859024   45.255127
4017 2025-12-31  34.201460   23.032949   45.044599


In [31]:
#initialize prophet for each pollutant
model_co = Prophet()
model_co.fit(delhi_co)

#future timestamps
future_co = model_co.make_future_dataframe(periods=365)

#predictions
forecast_co = model_co.predict(future_co)

#print results
print(forecast_co[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:32 - cmdstanpy - INFO - Chain [1] start processing
12:01:32 - cmdstanpy - INFO - Chain [1] done processing


             ds      yhat  yhat_lower  yhat_upper
4013 2025-12-27  4.165941    2.948972    5.400025
4014 2025-12-28  4.127683    3.013378    5.332738
4015 2025-12-29  4.553342    3.366296    5.612356
4016 2025-12-30  4.597987    3.491786    5.821866
4017 2025-12-31  4.573895    3.413871    5.706085


In [32]:
#initialize prophet for each pollutant
model_so2 = Prophet()
model_so2.fit(delhi_so2)

#future timestamps
future_so2 = model_so2.make_future_dataframe(periods=365)

#predictions
forecast_so2 = model_so2.predict(future_so2)

#print results
print(forecast_so2[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:33 - cmdstanpy - INFO - Chain [1] start processing
12:01:33 - cmdstanpy - INFO - Chain [1] done processing


             ds       yhat  yhat_lower  yhat_upper
4013 2025-12-27  41.090458   28.715413   53.057104
4014 2025-12-28  41.129096   29.847181   53.130520
4015 2025-12-29  45.663044   35.395678   57.272102
4016 2025-12-30  45.182344   33.680446   56.914726
4017 2025-12-31  45.519421   33.833942   57.505852


In [ ]:
#initialize prophet for each pollutant
model_o3 = Prophet()
model_o3.fit(delhi_o3)

#future timestamps
future_o3 = model_o3.make_future_dataframe(periods=365)

#predictions
forecast_o3 = model_o3.predict(future_o3)

#print results
print(forecast_o3[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail())

12:01:35 - cmdstanpy - INFO - Chain [1] start processing
12:01:35 - cmdstanpy - INFO - Chain [1] done processing


### Plotting the forecast and trend

In [ ]:
model_pm25.plot(forecast_pm25)
plt.title('PM2.5 Forcasted')
plt.show()

In [ ]:
model_pm25.plot_components(forecast_pm25)
plt.suptitle('PM2.5 Forecast Components')
plt.show()

In [ ]:
delhi_pm25.describe()

In [ ]:
forecast_pm25['yhat'].describe()

## Observation

Concentration of PM2.5 have a down trend from 2016 till 2024, it is maximum on weekdays throughout the year, and it is minimum during monsoon(July-Sept)

Similar pattern can be seen with other pollutants.

## Correlation between Weather and AQI

In [ ]:
delhi_df.columns

In [ ]:
delhi_df[['AQI','Max_Temperature_C', 'Min_Temperature_C', 'Avg_Temperature_C', 'Humidity_Percent', 'Rainfall_mm', 'Wind_Speed_kmh', 'Atmospheric_Pressure_hPa', 'Visibility_km']].corr()

In [ ]:
#visualising with heatmap
sns.heatmap(delhi_df[['AQI','Max_Temperature_C', 'Min_Temperature_C', 'Avg_Temperature_C', 'Humidity_Percent', 'Rainfall_mm', 'Wind_Speed_kmh', 'Atmospheric_Pressure_hPa', 'Visibility_km']].corr())

## Observation

AQI is highly affected by temparature, humidity and rainfall, but it is seen that it does not get much affected from atmospheric pressure.